src/stcal/ramp_fitting/gls_fit.py

```diff
@@ -141,120 +126,515 @@ def gls_ramp_fit(input_model, buffsize, save_opt, readnoise_model, gain_model, m
+    for the arguments needed for ols_ramp_fit_single.
+
+    ramp_data: RampData
+        The ramp data to be sliced.
+
+    save_opt : bool
+        Whether to return the optional output model
+
+    readnoise_2d : ndarray
+        The read noise of each pixel
+
+    gain_2d : ndarray
+        The gain of each pixel
+
+    max_num_cr: int
+        The maximum number of cosmic rays in a ramp.
+
+    number_slices: int
+        The number of slices to partition the data into for multiprocessing.
+
+    Return
+    ------
+    slices: list
+        The list of arguments for each processor for multiprocessing.
+    """
+    nrows = ramp_data.data.shape[2]
+    rslices = rows_per_slice(number_slices, nrows)
+    slices = []
+    start_row = 0
+    for k in range(len(rslices)):
+        ramp_slice = slice_ramp_data(ramp_data, start_row, rslices[k])
+        rnoise_slice = readnoise_2d[start_row:start_row + rslices[k], :].copy()
+        gain_slice = gain_2d[start_row:start_row + rslices[k], :].copy()
+        slices.insert(
+            k,
+            (ramp_slice, rnoise_slice, gain_slice, max_num_cr, save_opt))
+        start_row = start_row + rslices[k]
+
+    return slices, rslices
+
+
+def rows_per_slice(nslices, nrows):
+    """
+    Compute the number of rows per slice.
+
+    Parameters
+    ----------
+    nslices: int
+        The number of slices to partition the rows.
+
+    nrows: int
+        The number of rows to partition.
+
+    Return
+    ______
+    rslices: list
+        The number of rows for each slice.
+    """
+    quotient = nrows // nslices
+    remainder = nrows % nslices
+
+    no_inc = nslices - remainder
+    if remainder > 0:
+        # Ensure the number of rows per slice is no more than a
+        # difference of one.
+        first = [quotient + 1] * remainder
+        second = [quotient] * no_inc
+        rslices = first + second
+    else:
+        rslices = [quotient] * nslices
+
+    return rslices
+
+
+def slice_ramp_data(ramp_data, start_row, nrows):
+    """
+    Slices the ramp data by rows, where the arrays contain all rows in
+    [start_row, start_row+nrows).
+
+    Parameters
+    ----------
+    ramp_data: RampData
+        The ramp data to slice.
+
+    start_rows: int
+        The start row of the slice.
+
+    nrows: int
+        The number of rows in the slice.
+
+    Return
+    ------
+    ramp_data_slice: RampData
+        The slice of the ramp_data.
+    """
+    ramp_data_slice = ramp_fit_class.RampData()
+
+    # Slice data by row
+    data = ramp_data.data[:, :, start_row:start_row + nrows, :].copy()
+    err = ramp_data.err[:, :, start_row:start_row + nrows, :].copy()
+    groupdq = ramp_data.groupdq[:, :, start_row:start_row + nrows, :].copy()
+    pixeldq = ramp_data.pixeldq[start_row:start_row + nrows, :].copy()
+
+    ramp_data_slice.set_arrays(
+        data, err, groupdq, pixeldq, ramp_data.int_times)
+
+    # Carry over meta data.
+    ramp_data_slice.set_meta(
```

Comment:
The RampData class contains all information needed for ramp fitting process.  Global variables do not carry over, so must be reset on each process.  Because of this all meta data and DQ flags are saved in each RampData class, so each process will have that information.